contents/scalability/scalable-development.md

```diff
@@ -0,0 +1,63 @@
+
+However, due to Amdahl's law, each additional processor will give a less-than-proportionate speed up. Coupled with the exponential increase in price as the computing power of the CPU increases, this is typically not scalable in the long run.
+
+## Faster Algorithms
+
+If you have computationally-intensive functions, making them faster can help save a lot of time. For instance, if you want to look up files stored with a certain tag, a hashtable can be used to store files under commonly-searched tags, as opposed to a linear search. The difference between O(1) and O(n) increases as the number of files on a server.
+
+In certain scenarios, the runtime complexity can be reduced as a trade-off between time and accuracy. For example, if you want to figure out the type of content a user wants to see on a social network, rather than analyze the user's past history of posts liked, you can just sample the 100 most recent post liked by the user and analyze them.
+
+## Avoid Bottlenecks
+
+There are a few common bottlenecks that can be avoided such as database querying, reading or writing to files and slow communication across network. Different bottlenecks can be solved with their respective solutions. Use non-relational databases where feasible, as they are scalable. Cache recent or frequently-accessed data to reduce database querying and file I/O. Avoid large data movement across the network by keeping the data close to the computation.
+
+There are many program profilers available that you can use to find out the bottleneck in your server. By finding out where the server spends most computational time in, you can potentially save huge amounts of time.
+
+Performance bottlenecks ([resource](https://www.apicasystem.com/blog/5-common-performance-bottlenecks/))
+
+Non-Relational Database ([resource](http://www.jamesserra.com/archive/2015/08/relational-databases-vs-non-relational-databases/)) ([resource](https://www.pluralsight.com/blog/software-development/relational-non-relational-databases))
+
+## Get More Machines
+
+Buy more machines to host your server. This is called horizontal scaling. This method can help improve scalability by a huge factor, as each additional unit of processing power and RAM costs the same, no matter how many machines you buy. As this is a promising way to increase scalability, It is no surprise why large services such as Google and Facebook use thousands of computer servers just to provide service to all their users.
+
+However, to fully reap the benefits of multiple machines, you will need to learn various techniques to modify your server infrastructure. In most cases you cannot simply run multiple instances of the server and expect it to work.
+
+Here are some tips to get started with horizontal scaling:
```

Comment:
These are very userful resources! Maybe you can also categorise them into `services provider` (like GAE and companies that provide Hadoop ecosystem), versus guides to set up in-house infrastructure (MPI, server architecture)